src/Http/Routing/src/Internal/MapActionExpressionTreeBuilder.cs

```diff
@@ -0,0 +1,402 @@
+                                       ExecuteValueTaskOfTMethodInfo.MakeGenericMethod(typeArg),
+                                       methodCall,
+                                       TargetArg,
+                                       HttpContextParameter);
+                    }
+                }
+                else
+                {
+                    // TODO: Handle custom awaitables
+                    throw new NotSupportedException($"Unsupported return type: {method.ReturnType}");
+                }
+            }
+            else if (typeof(IResult).IsAssignableFrom(method.ReturnType))
+            {
+                body = Expression.Call(methodCall, ResultWriteResponseAsync, HttpContextParameter);
+            }
+            else if (method.ReturnType == typeof(string))
+            {
+                body = Expression.Call(StringResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+            else
+            {
+                body = Expression.Call(JsonResultWriteResponseAsync, HttpResponseExpr, methodCall, Expression.Constant(CancellationToken.None));
+            }
+
+            Func<object?, HttpContext, Task>? requestDelegate = null;
+
+            if (consumeBodyDirectly)
+            {
+                // We need to generate the code for reading from the body before calling into the delegate
+                var lambda = Expression.Lambda<Func<object?, HttpContext, object?, Task>>(body, TargetArg, HttpContextParameter, DeserializedBodyArg);
+                var invoker = lambda.Compile();
+                object? defaultBodyValue = null;
+
+                if (allowEmptyBody && bodyType!.IsValueType)
+                {
+                    defaultBodyValue = Activator.CreateInstance(bodyType);
+                }
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    object? bodyValue;
+
+                    if (allowEmptyBody && httpContext.Request.ContentLength == 0)
+                    {
+                        bodyValue = defaultBodyValue;
+                    }
+                    else
+                    {
+                        bodyValue = await httpContext.Request.ReadFromJsonAsync(bodyType!);
+                    }
+
+                    await invoker(target, httpContext, bodyValue);
+                };
+            }
+            else if (consumeBodyAsForm)
+            {
+                var lambda = Expression.Lambda<Func<object?, HttpContext, Task>>(body, TargetArg, HttpContextParameter);
+                var invoker = lambda.Compile();
+
+                requestDelegate = async (target, httpContext) =>
+                {
+                    // Generating async code would just be insane so if the method needs the form populate it here
+                    // so the within the method it's cached
+                    await httpContext.Request.ReadFormAsync();
```

Comment:
This needs to suppress form read errors like client disconnects and set a default 400 response status. Otherwise the logs (and our issues) fill up with useless exception stack traces.

The JSON code path should do something similar.